### ISSUE TITLE
simple convnet not working

### ISSUE THREAD

**Issue by JiahaoYao** [deepmind/dm-haiku#81]
Hi, 
I wonder why the following conv-net is not learning? is that because the gradient of the convolutional layer in Haiku? Does someone has the same experience? 

```python 


import itertools
import time

import haiku as hk
import jax
import jax.numpy as jnp
import numpy.random as npr
import optax
from examples import datasets
from jax import grad, jit, random
from jax.experimental import optimizers, stax
from jax.experimental.stax import (
    Dense,
    Flatten,
    GeneralConv,
    LogSoftmax,
    Relu,
    elementwise,
)


def net_fn(x) -> jnp.ndarray:
    """Standard LeNet-300-100 MLP network."""
    mlp = hk.Sequential(
        [
            hk.Conv2D(output_channels=16, kernel_shape=[5, 5], padding="SAME"),
            jax.nn.relu,
            hk.MaxPool(window_shape=[2, 2], strides=[2, 2], padding="VALID"),
            hk.Conv2D(output_channels=32, kernel_shape=[5, 5], padding="SAME"),
            jax.nn.relu,
            hk.MaxPool(window_shape=[2, 2], strides=[2, 2], padding="VALID"),
            hk.Flatten(),
            hk.Linear(10),
        ]
    )
    return mlp(x)


net = hk.without_apply_rng(hk.transform(net_fn))


def loss(params, batch):
    inputs, targets = batch
    preds = net.apply(params, inputs)
    return -jnp.mean(jnp.sum(preds * targets, axis=1))


def accuracy(params, batch):
    inputs, targets = batch
    target_class = jnp.argmax(targets, axis=1)
    predicted_class = jnp.argmax(net.apply(params, inputs), axis=1)
    return jnp.mean(predicted_class == target_class)


if __name__ == "__main__":

    step_size = 0.001
    num_epochs = 10
    batch_size = 128
    momentum_mass = 0.9

    train_images, train_labels, test_images, test_labels = datasets.mnist()

    num_train = train_images.shape[0]
    num_complete_batches, leftover = divmod(num_train, batch_size)
    num_batches = num_complete_batches + bool(leftover)

    def data_stream():
        rng = npr.RandomState(0)
        while True:
            perm = rng.permutation(num_train)
            for i in range(num_batches):
                batch_idx = perm[i * batch_size : (i + 1) * batch_size]
                yield train_images[batch_idx].reshape(-1, 1, 28, 28), train_labels[
                    batch_idx
                ]

    batches = data_stream()

    optimizer = optax.chain(
        optax.scale_by_adam(b1=0.9, b2=0.999, eps=1e-8), optax.scale(-step_size)
    )

    @jit
    def update(params, optimizer_state, batch):
        grads = grad(loss)(params, batch)
        optim_update, optimizer_state = optimizer.update(grads, optimizer_state, params)
        params = optax.apply_updates(params, optim_update)
        return params, optimizer_state

    params = net.init(jax.random.PRNGKey(42), next(batches)[0])

    # params = init_params
    optimizer_state = optimizer.init(params)

    itercount = itertools.count()

    print("\nStarting training...")
    for epoch in range(num_epochs):
        start_time = time.time()
        for _ in range(num_batches):
            params, optimizer_state = update(params, optimizer_state, next(batches))

        epoch_time = time.time() - start_time

        train_acc = accuracy(
            params, (train_images.reshape(-1, 1, 28, 28), train_labels)
        )
        test_acc = accuracy(params, (test_images.reshape(-1, 1, 28, 28), test_labels))
        print("Epoch {} in {:0.2f} sec".format(epoch, epoch_time))
        print("Training set accuracy {}".format(train_acc))
        print("Test set accuracy {}".format(test_acc))
```

The results are 
```
Starting training...
Epoch 0 in 1.73 sec
Training set accuracy 0.13341666758060455
Test set accuracy 0.14079999923706055
Epoch 1 in 0.34 sec
Training set accuracy 0.09730000048875809
Test set accuracy 0.10090000182390213
Epoch 2 in 0.32 sec
Training set accuracy 0.09966666996479034
Test set accuracy 0.10300000756978989
Epoch 3 in 0.33 sec
Training set accuracy 0.12701666355133057
Test set accuracy 0.12640000879764557
Epoch 4 in 0.34 sec
Training set accuracy 0.13825000822544098
Test set accuracy 0.13690000772476196
Epoch 5 in 0.37 sec
Training set accuracy 0.11961666494607925
Test set accuracy 0.1193000078201294
Epoch 6 in 0.34 sec
Training set accuracy 0.16324999928474426
Test set accuracy 0.1599000096321106
Epoch 7 in 0.33 sec
Training set accuracy 0.1628333330154419
Test set accuracy 0.16090001165866852
Epoch 8 in 0.38 sec
Training set accuracy 0.16438333690166473
Test set accuracy 0.1623000055551529
Epoch 9 in 0.33 sec
Training set accuracy 0.10546667128801346
Test set accuracy 0.10450000315904617
Epoch 10 in 0.35 sec
Training set accuracy 0.1001666709780693
Test set accuracy 0.09920000284910202
Epoch 11 in 0.34 sec
Training set accuracy 0.09881667047739029
Test set accuracy 0.09830000251531601
Epoch 12 in 0.32 sec
Training set accuracy 0.09871666878461838
Test set accuracy 0.09800000488758087
Epoch 13 in 0.33 sec
Training set accuracy 0.09881667047739029
Test set accuracy 0.09830000251531601
Epoch 14 in 0.34 sec
Training set accuracy 0.10068333148956299
Test set accuracy 0.10000000149011612
Epoch 15 in 0.35 sec
Training set accuracy 0.1003333330154419
Test set accuracy 0.0998000055551529
Epoch 16 in 0.37 sec
Training set accuracy 0.11216666549444199
Test set accuracy 0.11060000211000443
```

**Comment by tomhennigan** [comment]
I've made some edits to your example to get it converging https://colab.research.google.com/gist/tomhennigan/3ee83d39e1ad9f359ec122a6a26dad21/haiku-issue-81.ipynb.

```
Epoch 9 in 1.01 sec
Training set accuracy 0.9842666983604431
Test set accuracy 0.9717000722885132
```

The main changes here are:

- Changing `-jnp.mean(jnp.sum(preds * targets, axis=1))` to `-jnp.sum(targets * jax.nn.log_softmax(logits))` in `loss`.
- Passing NHWC input rather than NCHW (changing the reshapes from `1, 28, 28` to `28, 28, 1`).
- (Minor) I've used TensorFlow datasets since I was not sure where your `examples` module comes from. This required explicitly one_hot encoding the labels and casting images to f32.

It might be useful to look at our mnist example: https://github.com/deepmind/dm-haiku/blob/master/examples/mnist.py.
